tsdb/querier.go

```diff
@@ -530,25 +524,143 @@ func EmptySeriesSet() SeriesSet {
 	return emptySeriesSet
 }
 
-// mergedSeriesSet takes two series sets as a single series set. The input series sets
-// must be sorted and sequential in time, i.e. if they have the same label set,
-// the datapoints of a must be before the datapoints of b.
+// mergedSeriesSet returns a series sets slice as a single series set. The input series sets
+// must be sorted and sequential in time.
 type mergedSeriesSet struct {
+	all  []SeriesSet
+	buf  []SeriesSet // A buffer for keeping the order of SeriesSet slice during forwarding the SeriesSet.
+	ids  []int       // The indices of chosen SeriesSet for the current run.
+	done bool
+	err  error
+	cur  Series
+}
+
+func NewMergedSeriesSet(all []SeriesSet) SeriesSet {
+	if len(all) == 1 {
+		return all[0]
+	} else if len(all) == 2 {
+		return newBinaryMergedSeriesSet(all[0], all[1])
+	}
+	s := &mergedSeriesSet{all: all}
+	// Initialize first elements of all sets as Next() needs
+	// one element look-ahead.
+	s.nextAll()
+	if len(s.all) == 0 {
+		s.done = true
+	}
+
+	return s
+}
+
+func (s *mergedSeriesSet) At() Series {
+	return s.cur
+}
+
+func (s *mergedSeriesSet) Err() error {
+	return s.err
+}
+
+// This function is to call Next() for all SeriesSet.
+// Because the order of the SeriesSet slice will affect the results,
+// we need to use an buffer slice to hold the order.
+func (s *mergedSeriesSet) nextAll() {
+	s.buf = s.buf[:0]
+	for _, ss := range s.all {
+		if ss.Next() {
+			s.buf = append(s.buf, ss)
+		} else if ss.Err() != nil {
+			s.done = true
+			s.err = ss.Err()
+			break
+		}
+	}
+	s.all, s.buf = s.buf, s.all
+}
+
+// This function is to call Next() for the SeriesSet with the indices of s.ids.
```

Comment:
doesn't seem to describe the exact behavior of the function and the comment is an exact copy of `nextAll`